### PR TITLE
ARROW-16910: [C++] Add Equals method for FileFragment

### PR DESCRIPTION
**Pull request by vibhatha**
Adding Equals method for `FileFragment`

### CODE REVIEW

File: cpp/src/arrow/dataset/file_base.h [hunk 1]
```diff
@@ -114,6 +114,9 @@ class ARROW_DS_EXPORT FileSource {
   Result<std::shared_ptr<io::InputStream>> OpenCompressed(
       util::optional<Compression::type> compression = util::nullopt) const;
 
+  /// \brief equality comparison with another FileSource
+  bool Equals(const FileSource& other) const;
```
Comment (lidavidm) [inline cpp/src/arrow/dataset/file_base.h:118]: We may want to inherit from EqualityComparable https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/compare.h

File: cpp/src/arrow/dataset/test_util.h [hunk 2]
```diff
@@ -555,6 +555,21 @@ class FileFormatFixtureMixin : public ::testing::Test {
...
+    auto other_reader = this->GetRecordBatchReader(test_schema);
+    auto source = this->GetFileSource(reader.get());
+    auto other_source = this->GetFileSource(other_reader.get());
+
+    auto fragment = this->MakeFragment(*source);
+    auto other = this->MakeFragment(*other_source);
+
+    EXPECT_EQ(fragment->Equals(*other), false);
+    // TODO: Extend the test cases for Fragment
+    // when ARROW-16855 is merged.
```
Comment (lidavidm) [inline cpp/src/arrow/dataset/test_util.h:571]: Is that a TODO for the code here, or for the other PR?

File: cpp/src/arrow/dataset/file_base.cc [hunk 3]
```diff
@@ -89,6 +89,12 @@ Result<std::shared_ptr<io::InputStream>> FileSource::OpenCompressed(
   return io::CompressedInputStream::Make(codec.get(), std::move(file));
 }
 
+bool FileSource::Equals(const FileSource& other) const {
+  if (filesystem_ == NULLPTR) return false;
```
Comment (lidavidm) [inline cpp/src/arrow/dataset/file_base.cc:93]: This shouldn't be possible right?

Also `NULLPTR` is meant for headers, not source files

File: cpp/src/arrow/dataset/test_util.h (hunk 2 above)
Comment (vibhatha) [reply cpp/src/arrow/dataset/test_util.h:571]: Yes. I need to mention this in other PR as well.

File: cpp/src/arrow/dataset/file_base.h (hunk 1 above)
Comment (vibhatha) [reply cpp/src/arrow/dataset/file_base.h:118]: 👍

File: cpp/src/arrow/dataset/test_util.h (hunk 2 above)
Comment (vibhatha) [reply cpp/src/arrow/dataset/test_util.h:571]: Sure. I will move it to the other PR

File: cpp/src/arrow/dataset/file_base.cc (hunk 3 above)
Comment (vibhatha) [reply cpp/src/arrow/dataset/file_base.cc:93]: Mmm sorry didn’t catch this one clearly. Could you explain a bit?

File: cpp/src/arrow/dataset/file_base.h (hunk 1 above)
Comment (vibhatha) [reply cpp/src/arrow/dataset/file_base.h:118]: I added it, is it okay?

File: cpp/src/arrow/dataset/file_base.cc (hunk 3 above)
Comment (lidavidm) [reply cpp/src/arrow/dataset/file_base.cc:93]: Can filesystem_ be null? I would have assumed not given this is a FileFragment (and it should be validated in the constructor)

File: cpp/src/arrow/dataset/file_base.h (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/dataset/file_base.h:118]: I don't see the change for FileSource.

Comment (vibhatha) [reply cpp/src/arrow/dataset/file_base.h:118]: Ah I missed that. I will add this too.

File: cpp/src/arrow/dataset/file_base.cc [hunk 4]
```diff
@@ -89,6 +89,12 @@ Result<std::shared_ptr<io::InputStream>> FileSource::OpenCompressed(
   return io::CompressedInputStream::Make(codec.get(), std::move(file));
 }
 
+bool FileSource::Equals(const FileSource& other) const {
+  if (filesystem_ == NULLPTR) return false; // TODO: remove
```
Comment (lidavidm) [inline cpp/src/arrow/dataset/file_base.cc:93]: Please remove this then?

File: cpp/src/arrow/dataset/file_base.h [hunk 5]
```diff
@@ -46,7 +46,7 @@ namespace dataset {
 
 /// \brief The path and filesystem where an actual file is located or a buffer which can
 /// be read like a file
-class ARROW_DS_EXPORT FileSource {
+class ARROW_DS_EXPORT FileSource : public util::EqualityComparable<FileSource>{
```
Comment (lidavidm) [inline cpp/src/arrow/dataset/file_base.h:49]: I think you'll need to run the formatter

File: cpp/src/arrow/dataset/test_util.h [hunk 6]
```diff
@@ -555,6 +555,19 @@ class FileFormatFixtureMixin : public ::testing::Test {
...
+    auto test_schema = schema({field("f64", float64())});
+    auto reader = this->GetRecordBatchReader(test_schema);
+    auto other_reader = this->GetRecordBatchReader(test_schema);
+    auto source = this->GetFileSource(reader.get());
+    auto other_source = this->GetFileSource(other_reader.get());
+
+    auto fragment = this->MakeFragment(*source);
+    auto other = this->MakeFragment(*other_source);
+
+    EXPECT_EQ(fragment->Equals(*other), false);
```
Comment (lidavidm) [inline cpp/src/arrow/dataset/test_util.h:569]: Also check `fragment->Equals(*fragment)`?

Also use `EXPECT_TRUE` and `EXPECT_FALSE`

File: cpp/src/arrow/dataset/file_base.cc (hunk 4 above)
Comment (vibhatha) [reply cpp/src/arrow/dataset/file_base.cc:93]: Yeah that make sense. Let's do that.

**Comment by vibhatha** [comment]
cc @westonpace @pitrou Not sure what is wrong here. Do you have an idea?

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/44112333

File: cpp/src/arrow/dataset/file_base.cc [hunk 7]
```diff
@@ -89,6 +89,18 @@ Result<std::shared_ptr<io::InputStream>> FileSource::OpenCompressed(
   return io::CompressedInputStream::Make(codec.get(), std::move(file));
 }
 
+bool FileSource::Equals(const FileSource& other) const {
+  bool match_file_system = false;
+  if (filesystem_ == NULLPTR && other.filesystem_ == NULLPTR) {
+    match_file_system = true;
+  }
```
Comment (lidavidm) [inline cpp/src/arrow/dataset/file_base.cc:96]: don't do this, just

```cpp
bool match_file_system = (filesystem_ == nullptr && other.filesystem_ == nullptr) ||
                         (filesystem_ && other.filesystem_ && filesystem_->Equals(other.filesystem_);
```

File: cpp/src/arrow/dataset/file_base.cc [hunk 8]
```diff
@@ -89,6 +89,14 @@ Result<std::shared_ptr<io::InputStream>> FileSource::OpenCompressed(
   return io::CompressedInputStream::Make(codec.get(), std::move(file));
 }
 
+bool FileSource::Equals(const FileSource& other) const {
+  bool match_file_system =
+      (filesystem_ == nullptr && other.filesystem_ == nullptr) ||
+      (filesystem_ && other.filesystem_ && filesystem_->Equals(other.filesystem_));
+  return file_info_.Equals(other.file_info_) && match_file_system &&
```
Comment (lidavidm) [inline cpp/src/arrow/dataset/file_base.cc:96]: nit, but put `match_file_system` first to avoid unnecessarily evaluating `file_info_.Equals`?

Comment (vibhatha) [reply cpp/src/arrow/dataset/file_base.cc:96]: ah yes good catch.

**Review by lidavidm — APPROVED** [review]

**Pull request merged**